renku_notebooks/api/classes/k8s_client.py

```diff
@@ -183,6 +188,140 @@ def list_servers(self, label_selector: Optional[str] = None) -> List[Dict[str, A
+                return
+            raise
+        old_docker_config = json.loads(base64.b64decode(secret.data[".dockerconfigjson"]).decode())
+        hostname = next(iter(old_docker_config["auths"].keys()), None)
+        if not hostname:
+            raise ProgrammingError(
+                "Failed to refresh the access credentials in the image pull secret.",
+                detail="Please contact a Renku administrator.",
+            )
+        new_docker_config = {
+            "auths": {
+                hostname: {
+                    "Username": "oauth2",
+                    "Password": gitlab_token.access_token,
+                    "Email": old_docker_config["auths"][hostname]["Email"],
+                }
+            }
+        }
+        patch_path = "/data/.dockerconfigjson"
+        patch = [
+            {
+                "op": "replace",
+                "path": patch_path,
+                "value": base64.b64encode(json.dumps(new_docker_config).encode()).decode(),
+            }
+        ]
+        self._core_v1.patch_namespaced_secret(
+            secret_name,
+            self.namespace,
+            patch,
+        )
+
+    def patch_statefulset_tokens(
+        self, name: str, renku_tokens: RenkuTokens, gitlab_token: GitlabToken
+    ):
+        """Patch the Renku and Gitlab access tokens that are used in the session statefulset."""
+        try:
+            ss = self._apps_v1.read_namespaced_stateful_set(name, self.namespace)
+        except ApiException as err:
+            if err.status == 404:
+                # NOTE: It can happen potentially that another request or something else
+                # deleted the session as this request was going on, in this case we ignore
+                # the missing statefulset
+                return
+            raise
+        if (
+            len(ss.spec.template.spec.containers) < 3
+            or len(ss.spec.template.spec.init_containers) < 3
+        ):
+            raise ProgrammingError(
+                "The expected setup for a session was not found when trying to inject new tokens",
+                detail="Please contact a Renku administrator.",
+            )
+        git_proxy_container_index = 2
+        git_proxy_container = ss.spec.template.spec.containers[git_proxy_container_index]
+        git_init_container_index = 2
+        git_init_container = ss.spec.template.spec.init_containers[git_init_container_index]
+        patch = []
+        expires_at_env = find_env_var(git_proxy_container, "GITLAB_OAUTH_TOKEN_EXPIRES_AT")
+        gitlab_token_env = find_env_var(git_proxy_container, "GITLAB_OAUTH_TOKEN")
+        git_init_token_env = find_env_var(git_init_container, "GIT_CLONE_USER__OAUTH_TOKEN")
+        renku_access_token_env = find_env_var(git_proxy_container, "RENKU_ACCESS_TOKEN")
+        renku_refresh_token_env = find_env_var(git_proxy_container, "RENKU_REFRESH_TOKEN")
+        if not any(
```

Comment:
should this be `all`? now it'd pass if a single one is there

Reply:
Yes you are correct.